homeassistant/components/blue_current/__init__.py

```diff
@@ -157,34 +147,37 @@ def dispatch_grid_update_signal(self) -> None:
         """Dispatch a grid signal."""
         async_dispatcher_send(self.hass, f"{DOMAIN}_grid_update")
 
-    async def start_loop(self) -> None:
+    async def connect(self) -> None:
         """Start the receive loop."""
         try:
-            await self.client.start_loop(self.on_data)
-        except BlueCurrentException as err:
-            LOGGER.warning(
-                "Disconnected from the Blue Current websocket. Retrying to connect in background. %s",
-                err,
-            )
-
-            async_call_later(self.hass, SMALL_DELAY, self.reconnect)
-
-    async def reconnect(self, _event_time: datetime | None = None) -> None:
-        """Keep trying to reconnect to the websocket."""
-        try:
-            await self.connect(self.config.data[CONF_API_TOKEN])
-            LOGGER.debug("Reconnected to the Blue Current websocket")
-            self.hass.async_create_task(self.start_loop())
+            await self.client.connect(self.on_data)
         except RequestLimitReached:
-            self.available = False
-            async_call_later(
-                self.hass, self.client.get_next_reset_delta(), self.reconnect
+            LOGGER.warning(
+                "Request limit reached. reconnecting at 00:00 (Europe/Amsterdam)"
             )
+            self.on_disconnect()
+            delay = self.client.get_next_reset_delta()
+            await asyncio.sleep(delay.seconds)
+            await self.connect()
         except WebsocketError:
-            self.available = False
-            async_call_later(self.hass, LARGE_DELAY, self.reconnect)
+            LOGGER.debug("Disconnected, retrying in background")
+            self.on_disconnect()
+            await asyncio.sleep(DELAY)
+            await self.connect()
```

Comment:
```suggestion
    async def run_task(self) -> None:
        """Start the receive loop."""
        try:
            while True:
                try:
                    await self.client.connect(self.on_data)
                except RequestLimitReached:
                    LOGGER.warning(
                        "Request limit reached. reconnecting at 00:00 (Europe/Amsterdam)"
                    )
                    delay = self.client.get_next_reset_delta().seconds
                except WebsocketError:
                    LOGGER.debug("Disconnected, retrying in background")
                    delay = DELAY

                self._on_disconnect()
                await asyncio.sleep(delay)
        finally:
            await self._disconnect()
```

The functions `on_disconnect` and `disconnect` can be protected as they don't need to be called from outside